### ISSUE TITLE
Exception thrown: Cannot cast EmptyPage to ExtendPage

### ISSUE THREAD

**Issue by peters** [mbdavid/LiteDB#267]
This crash occured when attempting to insert a document on LiteDB 2.0 (latest version). I didn't have a PDB available at that time but I do now so I'll post a proper stacktrace and maybe provide a repro next time the program crashes.

```
 System.InvalidCastException: Unable to cast object of type 'LiteDB.EmptyPage' to type 'LiteDB.ExtendPage'.
   at LiteDB.PageService.GetPage[T](UInt32 pageID, Boolean setDirty)
   at LiteDB.PageService.<GetSeqPages>d__5`1.MoveNext()
   at LiteDB.DataService.ReadExtendData(UInt32 extendPageID)
   at LiteDB.DbEngine.<Find>d__9.MoveNext()
   at LiteDB.LiteCollection`1.<Find>d__34.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.OrderedEnumerable`1.<GetEnumerator>d__1.MoveNext()
```

/cc @mbdavid

**Comment by mbdavid** [comment]
Hi @peters, have you have this datafile to check? Can you reproduce this problem? It´s seans corrupted datafile

**Comment by oguimbal** [comment]
+1 ...  I encountered a similar exception:

    System.InvalidCastException: Impossible d'effectuer un cast d'un objet de type 'LiteDB.CollectionPage' en type 'LiteDB.HeaderPage'.
      à LiteDB.FileDiskService.ReadPage(UInt32 pageID)
      à LiteDB.PageService.GetPage[T](UInt32 pageID, Boolean setDirty)
      à LiteDB.DbEngine.GetCollectionPage(String name, Boolean addIfNotExits)
      à LiteDB.DbEngine.<Find>d__9.MoveNext()
      à LiteDB.LiteCollection`1.<Find>d__34.MoveNext()

Sorry, the stacktrace is in french  (it happened on one of my clients machine).

Happened only once so far, and the data file does not seem corrupt since the application continued to work as expected.

FYI my app accesses concurrurently to this db file (one instance of litedb per thread)

**Comment by mbdavid** [comment]
Hi @oguimbal, can you get lastest version from `master` branch? In my last commit I removed cache checkpoint that, in v2, has a bug that cause this error. Checkpoint is an write ahead that occurs when you use more than 5000 pages in memory. So, to avoid this I remove this feature.

If you use latest version and this problem occurs again, please tell me.

In v3 this checkpoint are back and this problem are fixed.

**Comment by oguimbal** [comment]
Thanks so much @mbdavid , i will do that.

Do you plan on releasing a 2.0.3 nuget package including this fix anytime soon ? Or should i build it on my own waiting for v3 to be released ?

**Comment by mbdavid** [comment]
I was waiting an OK from someone who has this error. I will release v2.0.3 in nuget before v3 (plans to release v3 only in december)

**Comment by mbdavid** [comment]
v3-beta released. If this problem continue in new version, re-open this issue. Thanks

**Comment by calvincotton** [comment]
Any word on this issue? I'm having a similar problem I think

**Comment by mbdavid** [comment]
Hi @calvincotton, this problem came back in new master version. I'm working on it. It's same as #399 It's all about single thread concurrency

**Comment by hardknoxni69a** [comment]
@mbdavid is this fixed yet as this is causing some issues in an app that we have your package in

**Comment by mbdavid** [comment]
Hi @hardknoxni69a, could you open new issue about this and point me on your code where are you using litedb?
